container-chains/templates/frontier/runtime/src/xcm_config.rs

```diff
@@ -244,3 +258,228 @@ impl cumulus_pallet_dmp_queue::Config for Runtime {
+    CheckingAccount,
+>;
+
+/// Multiplier used for dedicated `TakeFirstAssetTrader` with `ForeignAssets` instance.
+pub type AssetRateAsMultiplier = AssetFeeAsExistentialDepositMultiplier<
+    Runtime,
+    WeightToFee,
+    CustomConverter,
+    ForeignAssetsInstance,
+>;
+
+// TODO: move to https://github.com/paritytech/polkadot-sdk/pull/2903 once its merged
+pub struct CustomConverter;
+impl frame_support::traits::tokens::ConversionToAssetBalance<Balance, AssetId, Balance>
+    for CustomConverter
+{
+    type Error = ();
+    fn to_asset_balance(balance: Balance, asset_id: AssetId) -> Result<Balance, Self::Error> {
+        let rate = pallet_asset_rate::ConversionRateToNative::<Runtime>::get(asset_id).ok_or(())?;
+        Ok(sp_runtime::FixedU128::from_u32(1)
+            .checked_div(&rate)
+            .ok_or(())?
+            .saturating_mul_int(balance))
+    }
+}
+
+// TODO: this should probably move to somewhere in the polkadot-sdk repo
+pub struct NativeAssetReserve;
+impl frame_support::traits::ContainsPair<MultiAsset, MultiLocation> for NativeAssetReserve {
+    fn contains(asset: &MultiAsset, origin: &MultiLocation) -> bool {
+        log::trace!(target: "xcm::contains", "NativeAssetReserve asset: {:?}, origin: {:?}", asset, origin);
+        let reserve = if let Concrete(location) = &asset.id {
+            if location.parents == 0 && !matches!(location.first_interior(), Some(Parachain(_))) {
+                Some(MultiLocation::here())
+            } else {
+                location.chain_part()
+            }
+        } else {
+            None
+        };
+
+        if let Some(ref reserve) = reserve {
+            if reserve == origin {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+pub trait Parse {
+    /// Returns the "chain" location part. It could be parent, sibling
+    /// parachain, or child parachain.
+    fn chain_part(&self) -> Option<MultiLocation>;
+    /// Returns "non-chain" location part.
+    fn non_chain_part(&self) -> Option<MultiLocation>;
+}
+
+impl Parse for MultiLocation {
+    fn chain_part(&self) -> Option<MultiLocation> {
+        match (self.parents, self.first_interior()) {
+            // sibling parachain
+            (1, Some(Parachain(id))) => Some(MultiLocation::new(1, X1(Parachain(*id)))),
+            // parent
+            (1, _) => Some(MultiLocation::parent()),
+            // children parachain
+            (0, Some(Parachain(id))) => Some(MultiLocation::new(0, X1(Parachain(*id)))),
+            _ => None,
+        }
+    }
+
+    fn non_chain_part(&self) -> Option<MultiLocation> {
+        let mut junctions = *self.interior();
+        while matches!(junctions.first(), Some(Parachain(_))) {
+            let _ = junctions.take_first();
+        }
+
+        if junctions != Here {
+            Some(MultiLocation::new(0, junctions))
+        } else {
+            None
+        }
+    }
+}
+
+pub struct AssetFeeAsExistentialDepositMultiplier<
```

Comment:
Should dissappear once https://github.com/paritytech/polkadot-sdk/pull/3303 lands